clang/lib/Sema/SemaDeclAttr.cpp

```diff
@@ -2958,47 +2983,171 @@ static bool checkWorkGroupSizeValues(Sema &S, Decl *D, const ParsedAttr &Attr,
-          ReqdWorkGroupSizeAttr::CXX11_intel_reqd_work_group_size) {
-    WGSize[1] = ReqdWorkGroupSizeAttr::DefaultYDim;
-    WGSize[2] = ReqdWorkGroupSizeAttr::DefaultZDim;
-  } else if (!checkAttributeNumArgs(S, AL, 3))
+
+  Expr *XDimExpr = AL.getArgAsExpr(0);
+
+  // If no attribute argument is specified, set to default value '1'
+  // in ReqdWorkGroupSizeAttr for only with intel::reqd_work_group_size
+  // spelling.
+  Expr *YDimExpr =
+      AL.isArgExpr(1)
+          ? AL.getArgAsExpr(1)
+          : ((AL.getKind() == ParsedAttr::AT_ReqdWorkGroupSize &&
+              AL.getAttributeSpellingListIndex() ==
+                  ReqdWorkGroupSizeAttr::CXX11_intel_reqd_work_group_size)
+                 ? (IntegerLiteral::Create(S.Context, llvm::APInt(32, 1),
+                                           S.Context.IntTy, AL.getLoc()))
+                 : nullptr);
+
+  // If no attribute argument is specified, set to default value '1'
+  // in ReqdWorkGroupSizeAttr for only with intel::reqd_work_group_size
+  // spelling.
+  Expr *ZDimExpr =
+      AL.isArgExpr(2)
+          ? AL.getArgAsExpr(2)
+          : ((AL.getKind() == ParsedAttr::AT_ReqdWorkGroupSize &&
+              AL.getAttributeSpellingListIndex() ==
+                  ReqdWorkGroupSizeAttr::CXX11_intel_reqd_work_group_size)
+                 ? (IntegerLiteral::Create(S.Context, llvm::APInt(32, 1),
+                                           S.Context.IntTy, AL.getLoc()))
+                 : nullptr);
+
+  if ((AL.getKind() == ParsedAttr::AT_ReqdWorkGroupSize &&
+       AL.getAttributeSpellingListIndex() ==
+           ReqdWorkGroupSizeAttr::CXX11_cl_reqd_work_group_size) ||
+      (AL.getKind() == ParsedAttr::AT_SYCLIntelMaxWorkGroupSize)) {
+    if (!checkAttributeNumArgs(S, AL, 3))
+      return;
+  }
+
+  if (S.getLangOpts().SYCLIsDevice) {
+    std::swap(XDimExpr, ZDimExpr);
+  }
+
+  if (WorkGroupAttr *ExistingAttr = D->getAttr<WorkGroupAttr>()) {
+    Optional<llvm::APSInt> XDimVal =
+        XDimExpr->getIntegerConstantExpr(S.getASTContext());
+    Optional<llvm::APSInt> YDimVal =
+        YDimExpr->getIntegerConstantExpr(S.getASTContext());
+    Optional<llvm::APSInt> ZDimVal =
+        ZDimExpr->getIntegerConstantExpr(S.getASTContext());
+    Optional<llvm::APSInt> ExistingXDimVal =
+        ExistingAttr->getXDim()->getIntegerConstantExpr(S.getASTContext());
+    Optional<llvm::APSInt> ExistingYDimVal =
+        ExistingAttr->getYDim()->getIntegerConstantExpr(S.getASTContext());
+    Optional<llvm::APSInt> ExistingZDimVal =
+        ExistingAttr->getZDim()->getIntegerConstantExpr(S.getASTContext());
+
+    // Compare attribute arguments value and warn for a mismatch.
+    if (ExistingAttr &&
+        !((ExistingXDimVal->getExtValue() == XDimVal->getExtValue()) &&
+          (ExistingYDimVal->getExtValue() == YDimVal->getExtValue()) &&
+          (ExistingZDimVal->getExtValue() == ZDimVal->getExtValue()))) {
+      S.Diag(AL.getLoc(), diag::warn_duplicate_attribute) << AL;
```

Comment:
You should probably note the duplicate attribute location with a diagnostic note pointing at it.

Reply:
Done